In [1]:
#Librerias

import pandas as pd
import numpy as np

### **LECTURA DE DATOS**

In [2]:
df_general = pd.read_csv('data//general_data.csv')

# tamaño de dataframe general
df_general.shape

(8820, 24)

In [4]:
df_encuesta = pd.read_csv('data//employee_survey_data.csv')

# tamaño de dataframe encuesta
df_encuesta.shape

(8820, 5)

In [5]:
df_manager = pd.read_csv('data//manager_survey.csv')

# tamaño de dataframe manager
df_manager.shape

(8820, 4)

In [6]:
df_retiro= pd.read_csv('data//retirement_info.csv')

# tamaño de dataframe retirement
df_retiro.shape

(711, 5)

Respecto a la base de datos general, se puede observar que al analizarla hay algunas columnas que son candidatas a ser eliminadas. 
Columnas como Over18, StandardHours y EmployeeCount, al ser constantes no presentan comportamiento significativo.

In [7]:
df_general.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8820 entries, 0 to 8819
Data columns (total 24 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Age                      8820 non-null   int64  
 1   BusinessTravel           8820 non-null   object 
 2   Department               8820 non-null   object 
 3   DistanceFromHome         8820 non-null   int64  
 4   Education                8820 non-null   int64  
 5   EducationField           8820 non-null   object 
 6   EmployeeCount            8820 non-null   int64  
 7   EmployeeID               8820 non-null   int64  
 8   Gender                   8820 non-null   object 
 9   JobLevel                 8820 non-null   int64  
 10  JobRole                  8820 non-null   object 
 11  MaritalStatus            8820 non-null   object 
 12  MonthlyIncome            8820 non-null   int64  
 13  NumCompaniesWorked       8782 non-null   float64
 14  Over18                  

### **UNIÓN DE TABLAS**

In [8]:
#Renombrar columnna de fecha que aparece como InfoDate a Year
df_general = df_general.rename(columns={'InfoDate': 'Year'})

In [9]:
#revision de columnas
df_general

Age     BusinessTravel              Department  DistanceFromHome  \
0      51      Travel_Rarely                   Sales                 6   
1      31  Travel_Frequently  Research & Development                10   
2      32  Travel_Frequently  Research & Development                17   
3      38         Non-Travel  Research & Development                 2   
4      32      Travel_Rarely  Research & Development                10   
...   ...                ...                     ...               ...   
8815   42      Travel_Rarely  Research & Development                 5   
8816   29      Travel_Rarely  Research & Development                 2   
8817   25      Travel_Rarely  Research & Development                25   
8818   42      Travel_Rarely                   Sales                18   
8819   40      Travel_Rarely  Research & Development                28   

      Education EducationField  EmployeeCount  EmployeeID  Gender  JobLevel  \
0             2  Life Sciences              1           1  Female         1   
1             1  Life Sciences              1           2  Female         1   
2             4          Other              1           3    Male         4   
3             5  Life Sciences              1           4    Male         3   
4             1        Medical              1           5    Male         1   
...         ...            ...            ...         ...     ...       ...   
8815          4        Medical              1        4406  Female         1   
8816          4        Medical              1        4407    Male         1   
8817          2  Life Sciences              1        4408    Male         2   
8818          2        Medical              1        4409    Male         1   
8819          3        Medical              1        4410    Male         2   

      ... Over18 PercentSalaryHike  StandardHours  StockOptionLevel  \
0     ...      Y                11              8                 0   
1     ...      Y                23              8                 1   
2     ...      Y                15              8                 3   
3     ...      Y                11              8                 3   
4     ...      Y                12              8                 2   
...   ...    ...               ...            ...               ...   
8815  ...      Y                17              8                 1   
8816  ...      Y                15              8                 0   
8817  ...      Y                20              8                 0   
8818  ...      Y                14              8                 1   
8819  ...      Y                12              8                 0   

     TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
0                  1.0                      6               1   
1                  6.0                      3               5   
2                  5.0                      2               5   
3                 13.0                      5               8   
4                  9.0                      2               6   
...                ...                    ...             ...   
8815              10.0                      5               3   
8816              10.0                      2               3   
8817               5.0                      4               4   
8818              10.0                      2               9   
8819               NaN                      6              21   

      YearsSinceLastPromotion  YearsWithCurrManager        Year  
0                           0                     0  2015-12-31  
1                           1                     4  2015-12-31  
2                           0                     3  2015-12-31  
3                           7                     5  2015-12-31  
4                           0                     4  2015-12-31  
...                       ...                   ...         ...  
8815                        0                     2  2016-12-31  
8816  

In [10]:
#Renombrar columnna de fecha que aparece como DateSurvey a Year
df_encuesta = df_encuesta.rename(columns={'DateSurvey': 'Year'})

In [11]:
#Revision de columnas
df_encuesta

EmployeeID  EnvironmentSatisfaction  JobSatisfaction  WorkLifeBalance  \
0              1                      3.0              4.0              2.0   
1              2                      3.0              2.0              4.0   
2              3                      2.0              2.0              1.0   
3              4                      4.0              4.0              3.0   
4              5                      4.0              1.0              3.0   
...          ...                      ...              ...              ...   
8815        4406                      4.0              1.0              3.0   
8816        4407                      4.0              4.0              3.0   
8817        4408                      1.0              3.0              3.0   
8818        4409                      4.0              1.0              3.0   
8819        4410                      1.0              3.0              NaN   

            Year  
0     2015-12-31  
1     2015-12-31  
2     2015-12-31  
3     2015-12-31  
4     2015-12-31  
...          ...  
8815  2016-12-31  
8816  2016-12-31  
8817  2016-12-31  
8818  2016-12-31  
8819  2016-12-31  

[8820 rows x 5 columns]

In [12]:
#Renombrar columnna de fecha que aparece como SurveyDate a Year
df_manager = df_manager.rename(columns={'SurveyDate': 'Year'})

In [13]:
#Revision de columnas
df_manager

EmployeeID  JobInvolvement  PerformanceRating        Year
0              1               3                  3  2015-12-31
1              2               2                  4  2015-12-31
2              3               3                  3  2015-12-31
3              4               2                  3  2015-12-31
4              5               3                  3  2015-12-31
...          ...             ...                ...         ...
8815        4406               3                  3  2016-12-31
8816        4407               2                  3  2016-12-31
8817        4408               3                  4  2016-12-31
8818        4409               2                  3  2016-12-31
8819        4410               4                  3  2016-12-31

[8820 rows x 4 columns]

MERGED 1

En el primer merged de la base de datos se priorizó unir dos bases (General y Encuesta) que tuvieran la misma cantidad de observaciones (8820) para tener claro que el dataframe resultante debía tener 8820 de igual manera, solamente aumentando la cantidad de columnas. La union se realizo con EmployeeID y con Year para garantizar que sean los empleados del mismo año en y el mismo ID.

In [14]:
#Union de lo dataframes
df_merged1 = pd.merge(df_general, df_encuesta, on=['EmployeeID', 'Year'], how='outer')

In [15]:
#Revision de columnas
df_merged1

Age     BusinessTravel              Department  DistanceFromHome  \
0      51      Travel_Rarely                   Sales                 6   
1      31  Travel_Frequently  Research & Development                10   
2      32  Travel_Frequently  Research & Development                17   
3      38         Non-Travel  Research & Development                 2   
4      32      Travel_Rarely  Research & Development                10   
...   ...                ...                     ...               ...   
8815   42      Travel_Rarely  Research & Development                 5   
8816   29      Travel_Rarely  Research & Development                 2   
8817   25      Travel_Rarely  Research & Development                25   
8818   42      Travel_Rarely                   Sales                18   
8819   40      Travel_Rarely  Research & Development                28   

      Education EducationField  EmployeeCount  EmployeeID  Gender  JobLevel  \
0             2  Life Sciences              1           1  Female         1   
1             1  Life Sciences              1           2  Female         1   
2             4          Other              1           3    Male         4   
3             5  Life Sciences              1           4    Male         3   
4             1        Medical              1           5    Male         1   
...         ...            ...            ...         ...     ...       ...   
8815          4        Medical              1        4406  Female         1   
8816          4        Medical              1        4407    Male         1   
8817          2  Life Sciences              1        4408    Male         2   
8818          2        Medical              1        4409    Male         1   
8819          3        Medical              1        4410    Male         2   

      ... StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  \
0     ...                0               1.0                      6   
1     ...                1               6.0                      3   
2     ...                3               5.0                      2   
3     ...                3              13.0                      5   
4     ...                2               9.0                      2   
...   ...              ...               ...                    ...   
8815  ...                1              10.0                      5   
8816  ...                0              10.0                      2   
8817  ...                0               5.0                      4   
8818  ...                1              10.0                      2   
8819  ...                0               NaN                      6   

      YearsAtCompany YearsSinceLastPromotion  YearsWithCurrManager  \
0                  1                       0                     0   
1                  5                       1                     4   
2                  5                       0                     3   
3                  8                       7                     5   
4                  6                       0                     4   
...              ...                     ...                   ...   
8815               3                       0                     2   
8816               3                       0                     2   
8817               4                       1                     2   
8818               9                       7                     8   
8819              21                       3                     9   

            Year  EnvironmentSatisfaction  JobSatisfaction  WorkLifeBalance  
0     2015-12-31                      3.0              4.0              2.0  
1     2015-12-31                      3.0              2.0              4.0  
2     2015-12-31                      2.0              2.0              1.0  
3     2015-12-31                      4.0              4.0              3.0  
4     2015-12-31                      4.0              1.0              3.0  
...   

MERGED 2

En el segundo merged se decide unirlo con la base de datos de manager ya que tambien cuenta con 8820 observaciones y permite una revisión fácil para confirmar que no presente problemas. Además también se vuelven a unir según las columnas de EmployeeID y Year.

In [16]:
#creacion del segundo merged donde ya se han consolidado 3 bases de datos 
df_merged2 = pd.merge(df_merged1,df_manager, on=['EmployeeID','Year'],how='outer')

In [17]:
#Revision del segundo merged
df_merged2

Age     BusinessTravel              Department  DistanceFromHome  \
0      51      Travel_Rarely                   Sales                 6   
1      31  Travel_Frequently  Research & Development                10   
2      32  Travel_Frequently  Research & Development                17   
3      38         Non-Travel  Research & Development                 2   
4      32      Travel_Rarely  Research & Development                10   
...   ...                ...                     ...               ...   
8815   42      Travel_Rarely  Research & Development                 5   
8816   29      Travel_Rarely  Research & Development                 2   
8817   25      Travel_Rarely  Research & Development                25   
8818   42      Travel_Rarely                   Sales                18   
8819   40      Travel_Rarely  Research & Development                28   

      Education EducationField  EmployeeCount  EmployeeID  Gender  JobLevel  \
0             2  Life Sciences              1           1  Female         1   
1             1  Life Sciences              1           2  Female         1   
2             4          Other              1           3    Male         4   
3             5  Life Sciences              1           4    Male         3   
4             1        Medical              1           5    Male         1   
...         ...            ...            ...         ...     ...       ...   
8815          4        Medical              1        4406  Female         1   
8816          4        Medical              1        4407    Male         1   
8817          2  Life Sciences              1        4408    Male         2   
8818          2        Medical              1        4409    Male         1   
8819          3        Medical              1        4410    Male         2   

      ... TrainingTimesLastYear YearsAtCompany  YearsSinceLastPromotion  \
0     ...                     6              1                        0   
1     ...                     3              5                        1   
2     ...                     2              5                        0   
3     ...                     5              8                        7   
4     ...                     2              6                        0   
...   ...                   ...            ...                      ...   
8815  ...                     5              3                        0   
8816  ...                     2              3                        0   
8817  ...                     4              4                        1   
8818  ...                     2              9                        7   
8819  ...                     6             21                        3   

      YearsWithCurrManager        Year  EnvironmentSatisfaction  \
0                        0  2015-12-31                      3.0   
1                        4  2015-12-31                      3.0   
2                        3  2015-12-31                      2.0   
3                        5  2015-12-31                      4.0   
4                        4  2015-12-31                      4.0   
...                    ...         ...                      ...   
8815                     2  2016-12-31                      4.0   
8816                     2  2016-12-31                      4.0   
8817                     2  2016-12-31                      1.0   
8818                     8  2016-12-31                      4.0   
8819                     9  2016-12-31                      1.0   

      JobSatisfaction  WorkLifeBalance  JobInvolvement  PerformanceRating  
0                 4.0              2.0               3                  3  
1                 2.0              4.0               2                  4  
2                 2.0              1.0               3                  3  
3                 4.0              3.0               2                  3  
4                 1.0              3.0               3                  3  
...   

In [18]:
#Se renombra la columna EmployeeID a Col_ID y se pone en la primera columna del dataframe para facilitar la revisión 

col_ID = df_merged2.pop('EmployeeID')

df_merged2.insert(0, 'EmployeeID', col_ID)

In [19]:
#Se renombra la columna Year a Col_Year y se pone en la segunda columna del dataframe para facilitar la revisión 

col_Year = df_merged2.pop('Year')

df_merged2.insert(1, 'Year', col_Year)

In [20]:
#Revision del segundo merged
df_merged2

EmployeeID        Year  Age     BusinessTravel              Department  \
0              1  2015-12-31   51      Travel_Rarely                   Sales   
1              2  2015-12-31   31  Travel_Frequently  Research & Development   
2              3  2015-12-31   32  Travel_Frequently  Research & Development   
3              4  2015-12-31   38         Non-Travel  Research & Development   
4              5  2015-12-31   32      Travel_Rarely  Research & Development   
...          ...         ...  ...                ...                     ...   
8815        4406  2016-12-31   42      Travel_Rarely  Research & Development   
8816        4407  2016-12-31   29      Travel_Rarely  Research & Development   
8817        4408  2016-12-31   25      Travel_Rarely  Research & Development   
8818        4409  2016-12-31   42      Travel_Rarely                   Sales   
8819        4410  2016-12-31   40      Travel_Rarely  Research & Development   

      DistanceFromHome  Education EducationField  EmployeeCount  Gender  ...  \
0                    6          2  Life Sciences              1  Female  ...   
1                   10          1  Life Sciences              1  Female  ...   
2                   17          4          Other              1    Male  ...   
3                    2          5  Life Sciences              1    Male  ...   
4                   10          1        Medical              1    Male  ...   
...                ...        ...            ...            ...     ...  ...   
8815                 5          4        Medical              1  Female  ...   
8816                 2          4        Medical              1    Male  ...   
8817                25          2  Life Sciences              1    Male  ...   
8818                18          2        Medical              1    Male  ...   
8819                28          3        Medical              1    Male  ...   

      TotalWorkingYears TrainingTimesLastYear YearsAtCompany  \
0                   1.0                     6              1   
1                   6.0                     3              5   
2                   5.0                     2              5   
3                  13.0                     5              8   
4                   9.0                     2              6   
...                 ...                   ...            ...   
8815               10.0                     5              3   
8816               10.0                     2              3   
8817                5.0                     4              4   
8818               10.0                     2              9   
8819                NaN                     6             21   

      YearsSinceLastPromotion  YearsWithCurrManager EnvironmentSatisfaction  \
0                           0                     0                     3.0   
1                           1                     4                     3.0   
2                           0                     3                     2.0   
3                           7                     5                     4.0   
4                           0                     4                     4.0   
...                       ...                   ...                     ...   
8815                        0                     2                     4.0   
8816                        0                     2                     4.0   
8817                        1                     2                     1.0   
8818                        7                     8                     4.0   
8819                        3                     9                     1.0   

      JobSatisfaction  WorkLifeBalance  JobInvolvement  PerformanceRating  
0                 4.0              2.0               3                  3  
1                 2.0              4.0               2                  4  
2                 2.0              1.0               3                  3  
3                 4.0              3.0               2        

In [21]:
'''''
Se utiliza query para revisar los empleaados que se retiraron en el 2015 no deberían estar en 
la base de datos del 2016 porque ellos estarían en el 2014 (la cual no se tiene)
'''
df_merged2.query("EmployeeID == 2")

EmployeeID        Year  Age     BusinessTravel              Department  \
1           2  2015-12-31   31  Travel_Frequently  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  Gender  ...  \
1                10          1  Life Sciences              1  Female  ...   

   TotalWorkingYears TrainingTimesLastYear YearsAtCompany  \
1                6.0                     3              5   

   YearsSinceLastPromotion  YearsWithCurrManager EnvironmentSatisfaction  \
1                        1                     4                     3.0   

   JobSatisfaction  WorkLifeBalance  JobInvolvement  PerformanceRating  
1              2.0              4.0               2                  4  

[1 rows x 29 columns]

In [22]:
#Cambiar la columna Year a tipo fecha (date time) y se reemplazan los valores de datetime en la columna Year otra vez pero con el nuevo formato
df_merged2['Year'] = pd.to_datetime(df_merged2['Year'])

In [23]:
#Del formato anterior extraemos solo el año ya que es la variable que nos importa
df_merged2['Year'] = df_merged2['Year'].dt.year

In [24]:
#Revision de los cambios efectuados
df_merged2

EmployeeID  Year  Age     BusinessTravel              Department  \
0              1  2015   51      Travel_Rarely                   Sales   
1              2  2015   31  Travel_Frequently  Research & Development   
2              3  2015   32  Travel_Frequently  Research & Development   
3              4  2015   38         Non-Travel  Research & Development   
4              5  2015   32      Travel_Rarely  Research & Development   
...          ...   ...  ...                ...                     ...   
8815        4406  2016   42      Travel_Rarely  Research & Development   
8816        4407  2016   29      Travel_Rarely  Research & Development   
8817        4408  2016   25      Travel_Rarely  Research & Development   
8818        4409  2016   42      Travel_Rarely                   Sales   
8819        4410  2016   40      Travel_Rarely  Research & Development   

      DistanceFromHome  Education EducationField  EmployeeCount  Gender  ...  \
0                    6          2  Life Sciences              1  Female  ...   
1                   10          1  Life Sciences              1  Female  ...   
2                   17          4          Other              1    Male  ...   
3                    2          5  Life Sciences              1    Male  ...   
4                   10          1        Medical              1    Male  ...   
...                ...        ...            ...            ...     ...  ...   
8815                 5          4        Medical              1  Female  ...   
8816                 2          4        Medical              1    Male  ...   
8817                25          2  Life Sciences              1    Male  ...   
8818                18          2        Medical              1    Male  ...   
8819                28          3        Medical              1    Male  ...   

      TotalWorkingYears TrainingTimesLastYear YearsAtCompany  \
0                   1.0                     6              1   
1                   6.0                     3              5   
2                   5.0                     2              5   
3                  13.0                     5              8   
4                   9.0                     2              6   
...                 ...                   ...            ...   
8815               10.0                     5              3   
8816               10.0                     2              3   
8817                5.0                     4              4   
8818               10.0                     2              9   
8819                NaN                     6             21   

      YearsSinceLastPromotion  YearsWithCurrManager EnvironmentSatisfaction  \
0                           0                     0                     3.0   
1                           1                     4                     3.0   
2                           0                     3                     2.0   
3                           7                     5                     4.0   
4                           0                     4                     4.0   
...                       ...                   ...                     ...   
8815                        0                     2                     4.0   
8816                        0                     2                     4.0   
8817                        1                     2                     1.0   
8818                        7                     8                     4.0   
8819                        3                     9                     1.0   

      JobSatisfaction  WorkLifeBalance  JobInvolvement  PerformanceRating  
0                 4.0              2.0               3                  3  
1                 2.0              4.0               2                  4  
2                 2.0              1.0               3                  3  
3                 4.0              3.0               2                  3  
4                 1.0              3.0               3    

MERGED 3

In [25]:
df_retiro

EmployeeID Attrition retirementDate retirementType resignationReason
0             2       Yes     2016-10-15    Resignation            Others
1             7       Yes     2016-07-20    Resignation            Stress
2            14       Yes     2016-11-08    Resignation            Others
3            29       Yes     2016-10-15    Resignation            Others
4            31       Yes     2016-02-08    Resignation            Salary
..          ...       ...            ...            ...               ...
706        4382       Yes     2015-04-18    Resignation            Salary
707        4387       Yes     2015-09-16    Resignation            Others
708        4389       Yes     2015-05-24    Resignation            Others
709        4392       Yes     2015-02-26          Fired               NaN
710        4403       Yes     2015-01-26    Resignation            Stress

[711 rows x 5 columns]

In [26]:
#Se convierte la columna RTYear a tipo fecha (date time)
df_retiro['retirementDate'] = pd.to_datetime(df_retiro['retirementDate'])

In [27]:
#Del formato date time extraemos solo el año y la metemos en la nueva columna llamada Year
df_retiro['retirementDate'] = df_retiro['retirementDate'].dt.year

In [28]:
df_retiro

EmployeeID Attrition  retirementDate retirementType resignationReason
0             2       Yes            2016    Resignation            Others
1             7       Yes            2016    Resignation            Stress
2            14       Yes            2016    Resignation            Others
3            29       Yes            2016    Resignation            Others
4            31       Yes            2016    Resignation            Salary
..          ...       ...             ...            ...               ...
706        4382       Yes            2015    Resignation            Salary
707        4387       Yes            2015    Resignation            Others
708        4389       Yes            2015    Resignation            Others
709        4392       Yes            2015          Fired               NaN
710        4403       Yes            2015    Resignation            Stress

[711 rows x 5 columns]

In [29]:
#Filtramos solamente a un año (el 2015 no nos interesa de momento) y posteriormente será eliminado 
df_2015= df_retiro[df_retiro['retirementDate'] == 2015]

In [30]:
#Se crea una copia del dataframe de retiro y a esta nueva copia se le eliminaron los datos del 2015
df_retiro = df_retiro.drop(df_2015.index)

In [31]:
df_retiro

EmployeeID Attrition  retirementDate retirementType resignationReason
0             2       Yes            2016    Resignation            Others
1             7       Yes            2016    Resignation            Stress
2            14       Yes            2016    Resignation            Others
3            29       Yes            2016    Resignation            Others
4            31       Yes            2016    Resignation            Salary
..          ...       ...             ...            ...               ...
645        3973       Yes            2016    Resignation            Others
646        3974       Yes            2016    Resignation            Others
647        3980       Yes            2016    Resignation            Others
648        3981       Yes            2016    Resignation            Others
649        3995       Yes            2016    Resignation            Others

[650 rows x 5 columns]

In [32]:
# Eliminación de la fecha de retiro ya que no se tendrá en cuenta
df_retiro.drop('retirementDate', axis=1, inplace=True)

In [33]:
df_retiro

EmployeeID Attrition retirementType resignationReason
0             2       Yes    Resignation            Others
1             7       Yes    Resignation            Stress
2            14       Yes    Resignation            Others
3            29       Yes    Resignation            Others
4            31       Yes    Resignation            Salary
..          ...       ...            ...               ...
645        3973       Yes    Resignation            Others
646        3974       Yes    Resignation            Others
647        3980       Yes    Resignation            Others
648        3981       Yes    Resignation            Others
649        3995       Yes    Resignation            Others

[650 rows x 4 columns]

In [34]:
#Creacion del tercer merged, en el cual ya se tienen los datos que nos interesa del dataframe retiro. Adicionalmente se unen de acuerdo a EmployeeID
df_merged3 = pd.merge(df_merged2,df_retiro, on=['EmployeeID'],how='outer')

In [35]:
#DATAFRAME FINAL
df_merged3

EmployeeID  Year  Age     BusinessTravel              Department  \
0              1  2015   51      Travel_Rarely                   Sales   
1              1  2016   51      Travel_Rarely                   Sales   
2              2  2015   31  Travel_Frequently  Research & Development   
3              3  2015   32  Travel_Frequently  Research & Development   
4              3  2016   32  Travel_Frequently  Research & Development   
...          ...   ...  ...                ...                     ...   
8815        8383  2016   34  Travel_Frequently  Research & Development   
8816        8384  2016   37      Travel_Rarely  Research & Development   
8817        8390  2016   26      Travel_Rarely         Human Resources   
8818        8391  2016   46      Travel_Rarely  Research & Development   
8819        8405  2016   31      Travel_Rarely  Research & Development   

      DistanceFromHome  Education    EducationField  EmployeeCount  Gender  \
0                    6          2     Life Sciences              1  Female   
1                    6          2     Life Sciences              1  Female   
2                   10          1     Life Sciences              1  Female   
3                   17          4             Other              1    Male   
4                   17          4             Other              1    Male   
...                ...        ...               ...            ...     ...   
8815                 2          1     Life Sciences              1    Male   
8816                24          3     Life Sciences              1    Male   
8817                 2          1  Technical Degree              1  Female   
8818                10          3           Medical              1    Male   
8819                 2          3     Life Sciences              1    Male   

      ...  YearsSinceLastPromotion YearsWithCurrManager  \
0     ...                        0                    0   
1     ...                        0                    0   
2     ...                        1                    4   
3     ...                        0                    3   
4     ...                        0                    3   
...   ...                      ...                  ...   
8815  ...                        0                    7   
8816  ...                        0                    8   
8817  ...                        1                    4   
8818  ...                        3                    7   
8819  ...                        0                    0   

     EnvironmentSatisfaction  JobSatisfaction  WorkLifeBalance JobInvolvement  \
0                        3.0              4.0              2.0              3   
1                        3.0              4.0              2.0              3   
2                        3.0              2.0              4.0              2   
3                        2.0              2.0              1.0              3   
4                        2.0              2.0              1.0              3   
...                      ...              ...              ...            ...   
8815                     4.0              1.0              2.0              4   
8816                     4.0              1.0              1.0              3   
8817                     4.0              2.0              3.0              2   
8818                     4.0              2.0              4.0              3   
8819                     4.0              1.0              3.0              3   

      PerformanceRating  Attrition  retirementType  resignationReason  
0                     3        NaN             NaN                NaN  
1                     3        NaN             NaN                NaN  
2                     4        Yes     Resignation             Others  
3                     3        NaN             NaN                NaN  
4                     3        NaN             NaN                NaN  
...                 ...        ...             ...                

Finalmente, la base de datos consolidada es *'df_merged3'* la cual contiene todas las tablas, y a partir de esta se trabajará (se realizará la exploración de los datos y se alistará para el montaje de modelos)

In [36]:
# Visualizacion de los empleados que se retiraron
df_merged3[df_merged3['Attrition']=='Yes']

EmployeeID  Year  Age     BusinessTravel              Department  \
2              2  2015   31  Travel_Frequently  Research & Development   
11             7  2015   28      Travel_Rarely  Research & Development   
24            14  2015   47         Non-Travel  Research & Development   
53            29  2015   44  Travel_Frequently  Research & Development   
56            31  2015   26      Travel_Rarely  Research & Development   
...          ...   ...  ...                ...                     ...   
7299        3973  2015   34  Travel_Frequently  Research & Development   
7300        3974  2015   37      Travel_Rarely  Research & Development   
7311        3980  2015   26      Travel_Rarely         Human Resources   
7312        3981  2015   46      Travel_Rarely  Research & Development   
7339        3995  2015   31      Travel_Rarely  Research & Development   

      DistanceFromHome  Education    EducationField  EmployeeCount  Gender  \
2                   10          1     Life Sciences              1  Female   
11                  11          2           Medical              1    Male   
24                   1          1           Medical              1    Male   
53                   1          2           Medical              1    Male   
56                   4          3           Medical              1    Male   
...                ...        ...               ...            ...     ...   
7299                 2          1     Life Sciences              1    Male   
7300                24          3     Life Sciences              1    Male   
7311                 2          1  Technical Degree              1  Female   
7312                10          3           Medical              1    Male   
7339                 2          3     Life Sciences              1    Male   

      ...  YearsSinceLastPromotion YearsWithCurrManager  \
2     ...                        1                    4   
11    ...                        0                    0   
24    ...                        9                    9   
53    ...                        0                    0   
56    ...                        0                    2   
...   ...                      ...                  ...   
7299  ...                        0                    7   
7300  ...                        0                    8   
7311  ...                        1                    4   
7312  ...                        3                    7   
7339  ...                        0                    0   

     EnvironmentSatisfaction  JobSatisfaction  WorkLifeBalance JobInvolvement  \
2                        3.0              2.0              4.0              2   
11                       1.0              3.0              1.0              3   
24                       1.0              2.0              2.0              2   
53                       4.0              3.0              1.0              2   
56                       1.0              2.0              3.0              2   
...                      ...              ...              ...            ...   
7299                     4.0              1.0              2.0              4   
7300                     4.0              1.0              1.0              3   
7311                     4.0              2.0              3.0              2   
7312                     4.0              2.0              4.0              3   
7339                     4.0              1.0              3.0              3   

      PerformanceRating  Attrition  retirementType  resignationReason  
2                     4        Yes     Resignation             Others  
11                    4        Yes     Resignation             Stress  
24                    3        Yes     Resignation             Others  
53                    3        Yes     Resignation             Others  
56                    3        Yes     Resignation             Salary  
...                 ...        ...             ...                

### **IDENTIFICACIÓN E IMPUTACIÓN DE NULOS**

In [37]:
# Columnas que contienen valores nulos
columns_with_nulls = df_merged3.columns[df_merged3.isnull().any()]
columns_with_nulls

Index(['NumCompaniesWorked', 'TotalWorkingYears', 'EnvironmentSatisfaction',
       'JobSatisfaction', 'WorkLifeBalance', 'Attrition', 'retirementType',
       'resignationReason'],
      dtype='object')

In [38]:
df_merged3['WorkLifeBalance'].value_counts()

3.0    5320
2.0    2038
4.0     908
1.0     478
Name: WorkLifeBalance, dtype: int64

Se identifica que los datos nulos de algunas variables pueden ser reemplazadas con la mediana: 'NumCompaniesWorked', 'TotalWorkingYears', 'EnvironmentSatisfaction', 'JobSatisfaction' y 'WorkLifeBalance'.

In [39]:
var = ["NumCompaniesWorked", "TotalWorkingYears", "EnvironmentSatisfaction", "JobSatisfaction", "WorkLifeBalance"] #Definir las variables a tratar con la mediana

#Se utiliza una función para calcular la mediana para cada una de las variables seleccionadas anteriormente.

def imputar_mediana(df, columns):
    for col in columns:
        median_value = df[col].median() #Calcular la mediana
        df[col] = df[col].apply(lambda x: x if not pd.isnull(x) else int(median_value))
    return df

df_merged3 = imputar_mediana(df_merged3, var)

Y en el caso de las variables restantes (relacionadas con los datos de retiros) se entiende que las observaciones con datos nulos son los empleados que no se retiran. Por ende, la imputación de dichas datos debe ser por 'No aplica' o por valores de ese estilo.

In [40]:
# En el caso de 'Attrition' se reemplaza los nulos por 'No' dado que representa los empleados que no se retiran
df_merged3['Attrition'] = df_merged3['Attrition'].fillna('No')

In [41]:
# Además, con el objetivo de convertir la variable 'Attrition' en la variable objetivo más adelante, se reempaza 'Yes' por 1 y 'No' por 0 
df_merged3['Attrition'] = df_merged3['Attrition'].replace({'Yes': 1, 'No': 0})

In [42]:
# En el caso de las variables restantes ('retirementType', 'resignationReason') se reemplaza los nulos por 'Not Applicable'
df_merged3['retirementType'] = df_merged3['retirementType'].fillna('Not Applicable')
df_merged3['resignationReason'] = df_merged3['resignationReason'].fillna('Not Applicable')

In [43]:
df_merged3.columns

Index(['EmployeeID', 'Year', 'Age', 'BusinessTravel', 'Department',
       'DistanceFromHome', 'Education', 'EducationField', 'EmployeeCount',
       'Gender', 'JobLevel', 'JobRole', 'MaritalStatus', 'MonthlyIncome',
       'NumCompaniesWorked', 'Over18', 'PercentSalaryHike', 'StandardHours',
       'StockOptionLevel', 'TotalWorkingYears', 'TrainingTimesLastYear',
       'YearsAtCompany', 'YearsSinceLastPromotion', 'YearsWithCurrManager',
       'EnvironmentSatisfaction', 'JobSatisfaction', 'WorkLifeBalance',
       'JobInvolvement', 'PerformanceRating', 'Attrition', 'retirementType',
       'resignationReason'],
      dtype='object')

In [44]:
df_merged3.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8820 entries, 0 to 8819
Data columns (total 32 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   EmployeeID               8820 non-null   int64  
 1   Year                     8820 non-null   int64  
 2   Age                      8820 non-null   int64  
 3   BusinessTravel           8820 non-null   object 
 4   Department               8820 non-null   object 
 5   DistanceFromHome         8820 non-null   int64  
 6   Education                8820 non-null   int64  
 7   EducationField           8820 non-null   object 
 8   EmployeeCount            8820 non-null   int64  
 9   Gender                   8820 non-null   object 
 10  JobLevel                 8820 non-null   int64  
 11  JobRole                  8820 non-null   object 
 12  MaritalStatus            8820 non-null   object 
 13  MonthlyIncome            8820 non-null   int64  
 14  NumCompaniesWorked      

In [45]:
# Se realiza el cambio de tipología de algunas variables
def convertir_int(df, columns):
    for col in columns:
        df[col] = df[col].astype(int)
    return df

df_merged3 = convertir_int(df_merged3, var)  

In [47]:
df_merged3.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8820 entries, 0 to 8819
Data columns (total 32 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   EmployeeID               8820 non-null   int64 
 1   Year                     8820 non-null   int64 
 2   Age                      8820 non-null   int64 
 3   BusinessTravel           8820 non-null   object
 4   Department               8820 non-null   object
 5   DistanceFromHome         8820 non-null   int64 
 6   Education                8820 non-null   int64 
 7   EducationField           8820 non-null   object
 8   EmployeeCount            8820 non-null   int64 
 9   Gender                   8820 non-null   object
 10  JobLevel                 8820 non-null   int64 
 11  JobRole                  8820 non-null   object
 12  MaritalStatus            8820 non-null   object
 13  MonthlyIncome            8820 non-null   int64 
 14  NumCompaniesWorked       8820 non-null  

In [48]:
df_merged3.head(8)

EmployeeID  Year  Age     BusinessTravel              Department  \
0           1  2015   51      Travel_Rarely                   Sales   
1           1  2016   51      Travel_Rarely                   Sales   
2           2  2015   31  Travel_Frequently  Research & Development   
3           3  2015   32  Travel_Frequently  Research & Development   
4           3  2016   32  Travel_Frequently  Research & Development   
5           4  2015   38         Non-Travel  Research & Development   
6           4  2016   38         Non-Travel  Research & Development   
7           5  2015   32      Travel_Rarely  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  Gender  ...  \
0                 6          2  Life Sciences              1  Female  ...   
1                 6          2  Life Sciences              1  Female  ...   
2                10          1  Life Sciences              1  Female  ...   
3                17          4          Other              1    Male  ...   
4                17          4          Other              1    Male  ...   
5                 2          5  Life Sciences              1    Male  ...   
6                 2          5  Life Sciences              1    Male  ...   
7                10          1        Medical              1    Male  ...   

   YearsSinceLastPromotion YearsWithCurrManager EnvironmentSatisfaction  \
0                        0                    0                       3   
1                        0                    0                       3   
2                        1                    4                       3   
3                        0                    3                       2   
4                        0                    3                       2   
5                        7                    5                       4   
6                        7                    5                       4   
7                        0                    4                       4   

   JobSatisfaction  WorkLifeBalance JobInvolvement  PerformanceRating  \
0                4                2              3                  3   
1                4                2              3                  3   
2                2                4              2                  4   
3                2                1              3                  3   
4                2                1              3                  3   
5                4                3              2                  3   
6                4                3              2                  3   
7                1                3              3                  3   

   Attrition  retirementType  resignationReason  
0          0  Not Applicable     Not Applicable  
1          0  Not Applicable     Not Applicable  
2          1     Resignation             Others  
3          0  Not Applicable     Not Applicable  
4          0  Not Applicable     Not Applicable  
5          0  Not Applicable     Not Applicable  
6          0  Not Applicable     Not Applicable  
7          0  Not Applicable     Not Applicable  

[8 rows x 32 columns]

In [49]:
df_merged3 = df_merged3.sort_values(by=['Year', 'EmployeeID']) #Ordenar y filtrar la base de datos

In [50]:
df_merged3['JobRole'].value_counts()

Sales Executive              1956
Research Scientist           1752
Laboratory Technician        1554
Manufacturing Director        870
Healthcare Representative     786
Manager                       612
Sales Representative          498
Research Director             480
Human Resources               312
Name: JobRole, dtype: int64

### Datos necesarios para dashboard de exploración

A continuación se realizan algunos filtros necesarios para consolidar algunos datos a utilizar en la visualización de la exploración, pero su única finalidad es esa, no se utilizan en otro procedimiento.

In [51]:
# Datos completos de solo los empleados que se retiran a usar en dashboard
df_only_retired = df_merged3[df_merged3['Attrition']==1]
df_only_retired

EmployeeID  Year  Age     BusinessTravel              Department  \
2              2  2015   31  Travel_Frequently  Research & Development   
11             7  2015   28      Travel_Rarely  Research & Development   
24            14  2015   47         Non-Travel  Research & Development   
53            29  2015   44  Travel_Frequently  Research & Development   
56            31  2015   26      Travel_Rarely  Research & Development   
...          ...   ...  ...                ...                     ...   
7299        3973  2015   34  Travel_Frequently  Research & Development   
7300        3974  2015   37      Travel_Rarely  Research & Development   
7311        3980  2015   26      Travel_Rarely         Human Resources   
7312        3981  2015   46      Travel_Rarely  Research & Development   
7339        3995  2015   31      Travel_Rarely  Research & Development   

      DistanceFromHome  Education    EducationField  EmployeeCount  Gender  \
2                   10          1     Life Sciences              1  Female   
11                  11          2           Medical              1    Male   
24                   1          1           Medical              1    Male   
53                   1          2           Medical              1    Male   
56                   4          3           Medical              1    Male   
...                ...        ...               ...            ...     ...   
7299                 2          1     Life Sciences              1    Male   
7300                24          3     Life Sciences              1    Male   
7311                 2          1  Technical Degree              1  Female   
7312                10          3           Medical              1    Male   
7339                 2          3     Life Sciences              1    Male   

      ...  YearsSinceLastPromotion YearsWithCurrManager  \
2     ...                        1                    4   
11    ...                        0                    0   
24    ...                        9                    9   
53    ...                        0                    0   
56    ...                        0                    2   
...   ...                      ...                  ...   
7299  ...                        0                    7   
7300  ...                        0                    8   
7311  ...                        1                    4   
7312  ...                        3                    7   
7339  ...                        0                    0   

     EnvironmentSatisfaction  JobSatisfaction  WorkLifeBalance JobInvolvement  \
2                          3                2                4              2   
11                         1                3                1              3   
24                         1                2                2              2   
53                         4                3                1              2   
56                         1                2                3              2   
...                      ...              ...              ...            ...   
7299                       4                1                2              4   
7300                       4                1                1              3   
7311                       4                2                3              2   
7312                       4                2                4              3   
7339                       4                1                3              3   

      PerformanceRating  Attrition  retirementType  resignationReason  
2                     4          1     Resignation             Others  
11                    4          1     Resignation             Stress  
24                    3          1     Resignation             Others  
53                    3          1     Resignation             Others  
56                    3          1     Resignation             Salary  
...                 ...        ...             ...                

In [ ]:
#df_only_retired.to_csv('df_only_retired.csv', index=False)

In [52]:
# Moviendo la columna 'Attrition' 2 posiciones a la derecha
cols = df_merged3.columns.tolist()
cols.insert(cols.index('Attrition')+2, cols.pop(cols.index('Attrition')))
df_merged3 = df_merged3[cols]

In [53]:
df_merged3.head(8)

EmployeeID  Year  Age     BusinessTravel              Department  \
0            1  2015   51      Travel_Rarely                   Sales   
2            2  2015   31  Travel_Frequently  Research & Development   
3            3  2015   32  Travel_Frequently  Research & Development   
5            4  2015   38         Non-Travel  Research & Development   
7            5  2015   32      Travel_Rarely  Research & Development   
9            6  2015   46      Travel_Rarely  Research & Development   
11           7  2015   28      Travel_Rarely  Research & Development   
12           8  2015   29      Travel_Rarely  Research & Development   

    DistanceFromHome  Education EducationField  EmployeeCount  Gender  ...  \
0                  6          2  Life Sciences              1  Female  ...   
2                 10          1  Life Sciences              1  Female  ...   
3                 17          4          Other              1    Male  ...   
5                  2          5  Life Sciences              1    Male  ...   
7                 10          1        Medical              1    Male  ...   
9                  8          3  Life Sciences              1  Female  ...   
11                11          2        Medical              1    Male  ...   
12                18          3  Life Sciences              1    Male  ...   

    YearsSinceLastPromotion YearsWithCurrManager EnvironmentSatisfaction  \
0                         0                    0                       3   
2                         1                    4                       3   
3                         0                    3                       2   
5                         7                    5                       4   
7                         0                    4                       4   
9                         7                    7                       3   
11                        0                    0                       1   
12                        0                    0                       1   

    JobSatisfaction  WorkLifeBalance JobInvolvement  PerformanceRating  \
0                 4                2              3                  3   
2                 2                4              2                  4   
3                 2                1              3                  3   
5                 4                3              2                  3   
7                 1                3              3                  3   
9                 2                2              3                  3   
11                3                1              3                  4   
12                2                3              3                  4   

    retirementType  resignationReason  Attrition  
0   Not Applicable     Not Applicable          0  
2      Resignation             Others          1  
3   Not Applicable     Not Applicable          0  
5   Not Applicable     Not Applicable          0  
7   Not Applicable     Not Applicable          0  
9   Not Applicable     Not Applicable          0  
11     Resignation             Stress          1  
12  Not Applicable     Not Applicable          0  

[8 rows x 32 columns]

In [48]:
#df_entire_data.to_csv('entire_data.csv', index=False)

Hasta acá llegan los filtros de datos para visualización. Ahora, se continua con el preprocesamiento restante.

### Continuación del preprocesamiento

In [54]:
df_merged3.head()

EmployeeID  Year  Age     BusinessTravel              Department  \
0           1  2015   51      Travel_Rarely                   Sales   
2           2  2015   31  Travel_Frequently  Research & Development   
3           3  2015   32  Travel_Frequently  Research & Development   
5           4  2015   38         Non-Travel  Research & Development   
7           5  2015   32      Travel_Rarely  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  Gender  ...  \
0                 6          2  Life Sciences              1  Female  ...   
2                10          1  Life Sciences              1  Female  ...   
3                17          4          Other              1    Male  ...   
5                 2          5  Life Sciences              1    Male  ...   
7                10          1        Medical              1    Male  ...   

   YearsSinceLastPromotion YearsWithCurrManager EnvironmentSatisfaction  \
0                        0                    0                       3   
2                        1                    4                       3   
3                        0                    3                       2   
5                        7                    5                       4   
7                        0                    4                       4   

   JobSatisfaction  WorkLifeBalance JobInvolvement  PerformanceRating  \
0                4                2              3                  3   
2                2                4              2                  4   
3                2                1              3                  3   
5                4                3              2                  3   
7                1                3              3                  3   

   retirementType  resignationReason  Attrition  
0  Not Applicable     Not Applicable          0  
2     Resignation             Others          1  
3  Not Applicable     Not Applicable          0  
5  Not Applicable     Not Applicable          0  
7  Not Applicable     Not Applicable          0  

[5 rows x 32 columns]

### **ELIMINACIÓN DE VARIABLES**

In [55]:
df_merged3['Year'].value_counts()

2015    4410
2016    4410
Name: Year, dtype: int64

In [56]:
df_merged3[df_merged3['EmployeeID']==4410]

EmployeeID  Year  Age BusinessTravel              Department  \
8168        4410  2015   40  Travel_Rarely  Research & Development   
8169        4410  2016   40  Travel_Rarely  Research & Development   

      DistanceFromHome  Education EducationField  EmployeeCount Gender  ...  \
8168                28          3        Medical              1   Male  ...   
8169                28          3        Medical              1   Male  ...   

      YearsSinceLastPromotion YearsWithCurrManager EnvironmentSatisfaction  \
8168                        3                    9                       1   
8169                        3                    9                       1   

      JobSatisfaction  WorkLifeBalance JobInvolvement  PerformanceRating  \
8168                3                3              4                  3   
8169                3                3              4                  3   

      retirementType  resignationReason  Attrition  
8168  Not Applicable     Not Applicable          0  
8169  Not Applicable     Not Applicable          0  

[2 rows x 32 columns]

In [57]:
df_merged3.columns

Index(['EmployeeID', 'Year', 'Age', 'BusinessTravel', 'Department',
       'DistanceFromHome', 'Education', 'EducationField', 'EmployeeCount',
       'Gender', 'JobLevel', 'JobRole', 'MaritalStatus', 'MonthlyIncome',
       'NumCompaniesWorked', 'Over18', 'PercentSalaryHike', 'StandardHours',
       'StockOptionLevel', 'TotalWorkingYears', 'TrainingTimesLastYear',
       'YearsAtCompany', 'YearsSinceLastPromotion', 'YearsWithCurrManager',
       'EnvironmentSatisfaction', 'JobSatisfaction', 'WorkLifeBalance',
       'JobInvolvement', 'PerformanceRating', 'retirementType',
       'resignationReason', 'Attrition'],
      dtype='object')

In [58]:
df_merged3['EmployeeCount'].value_counts()

1    8820
Name: EmployeeCount, dtype: int64

In [59]:
df_merged3['Over18'].value_counts()

Y    8820
Name: Over18, dtype: int64

In [60]:
df_merged3['StandardHours'].value_counts()

8    8820
Name: StandardHours, dtype: int64

Se identifica que las variables *'EmployeeCount'*, *'Over18'* y *'StandardHours'* solo cuentan con un valor, por ende se decide eliminarlas del dataframe.

In [61]:
# Eliminación de columnas
df_merged3 = df_merged3.drop('EmployeeCount', axis=1)
df_merged3 = df_merged3.drop('Over18', axis=1)
df_merged3 = df_merged3.drop('StandardHours', axis=1)

In [62]:
df_merged3.head()

EmployeeID  Year  Age     BusinessTravel              Department  \
0           1  2015   51      Travel_Rarely                   Sales   
2           2  2015   31  Travel_Frequently  Research & Development   
3           3  2015   32  Travel_Frequently  Research & Development   
5           4  2015   38         Non-Travel  Research & Development   
7           5  2015   32      Travel_Rarely  Research & Development   

   DistanceFromHome  Education EducationField  Gender  JobLevel  ...  \
0                 6          2  Life Sciences  Female         1  ...   
2                10          1  Life Sciences  Female         1  ...   
3                17          4          Other    Male         4  ...   
5                 2          5  Life Sciences    Male         3  ...   
7                10          1        Medical    Male         1  ...   

  YearsSinceLastPromotion YearsWithCurrManager  EnvironmentSatisfaction  \
0                       0                    0                        3   
2                       1                    4                        3   
3                       0                    3                        2   
5                       7                    5                        4   
7                       0                    4                        4   

   JobSatisfaction  WorkLifeBalance  JobInvolvement  PerformanceRating  \
0                4                2               3                  3   
2                2                4               2                  4   
3                2                1               3                  3   
5                4                3               2                  3   
7                1                3               3                  3   

   retirementType  resignationReason  Attrition  
0  Not Applicable     Not Applicable          0  
2     Resignation             Others          1  
3  Not Applicable     Not Applicable          0  
5  Not Applicable     Not Applicable          0  
7  Not Applicable     Not Applicable          0  

[5 rows x 29 columns]

In [63]:
df_merged3.columns

Index(['EmployeeID', 'Year', 'Age', 'BusinessTravel', 'Department',
       'DistanceFromHome', 'Education', 'EducationField', 'Gender', 'JobLevel',
       'JobRole', 'MaritalStatus', 'MonthlyIncome', 'NumCompaniesWorked',
       'PercentSalaryHike', 'StockOptionLevel', 'TotalWorkingYears',
       'TrainingTimesLastYear', 'YearsAtCompany', 'YearsSinceLastPromotion',
       'YearsWithCurrManager', 'EnvironmentSatisfaction', 'JobSatisfaction',
       'WorkLifeBalance', 'JobInvolvement', 'PerformanceRating',
       'retirementType', 'resignationReason', 'Attrition'],
      dtype='object')

In [64]:
df_merged3['JobRole'].value_counts()

Sales Executive              1956
Research Scientist           1752
Laboratory Technician        1554
Manufacturing Director        870
Healthcare Representative     786
Manager                       612
Sales Representative          498
Research Director             480
Human Resources               312
Name: JobRole, dtype: int64

In [65]:
df_merged3['Department'].value_counts()

Research & Development    5766
Sales                     2676
Human Resources            378
Name: Department, dtype: int64

Por otro lado, se identifica que la variable *'JobRole'* y *'Department'* ofrecen información similar. Así mismo, *'JobRole* presenta 9 categorías diferentes, lo que aumentaría la dimensión del dataframe una vez se lleve a formato dummies, por ende, se decide eliminar la variabale *'JobRole'* del dataframe. 

In [66]:
df_merged3 = df_merged3.drop('JobRole', axis=1)

In [67]:
df_merged3.head()

EmployeeID  Year  Age     BusinessTravel              Department  \
0           1  2015   51      Travel_Rarely                   Sales   
2           2  2015   31  Travel_Frequently  Research & Development   
3           3  2015   32  Travel_Frequently  Research & Development   
5           4  2015   38         Non-Travel  Research & Development   
7           5  2015   32      Travel_Rarely  Research & Development   

   DistanceFromHome  Education EducationField  Gender  JobLevel  ...  \
0                 6          2  Life Sciences  Female         1  ...   
2                10          1  Life Sciences  Female         1  ...   
3                17          4          Other    Male         4  ...   
5                 2          5  Life Sciences    Male         3  ...   
7                10          1        Medical    Male         1  ...   

  YearsSinceLastPromotion  YearsWithCurrManager  EnvironmentSatisfaction  \
0                       0                     0                        3   
2                       1                     4                        3   
3                       0                     3                        2   
5                       7                     5                        4   
7                       0                     4                        4   

   JobSatisfaction  WorkLifeBalance  JobInvolvement  PerformanceRating  \
0                4                2               3                  3   
2                2                4               2                  4   
3                2                1               3                  3   
5                4                3               2                  3   
7                1                3               3                  3   

   retirementType  resignationReason  Attrition  
0  Not Applicable     Not Applicable          0  
2     Resignation             Others          1  
3  Not Applicable     Not Applicable          0  
5  Not Applicable     Not Applicable          0  
7  Not Applicable     Not Applicable          0  

[5 rows x 28 columns]

In [68]:
df_agrupado = df_merged3.groupby('Attrition')['Gender'].value_counts()
df_agrupado

Attrition  Gender
0          Male      4886
           Female    3284
1          Male       406
           Female     244
Name: Gender, dtype: int64

Se evidencia que en el caso de la variable *'Gender'*, la proporción de hombres y mujeres es proporcional independientemente de la variable objetivo, por ende, esta variable no ofrece explicabilidad relevante, y por eso se elimina.

In [69]:
df_merged3 = df_merged3.drop('Gender', axis=1)

In [70]:
df_agrupado2 = df_merged3.groupby('Attrition')['BusinessTravel'].value_counts()
df_agrupado2

Attrition  BusinessTravel   
0          Travel_Rarely        5836
           Travel_Frequently    1469
           Non-Travel            865
1          Travel_Rarely         422
           Travel_Frequently     193
           Non-Travel             35
Name: BusinessTravel, dtype: int64

In [71]:
df_merged3['PerformanceRating'].value_counts()

3    7464
4    1356
Name: PerformanceRating, dtype: int64

Finalmente, se encuentra que la variable *'PerformanceRating'* solo posee 2 valores (3 y 4), lo que básicamente no nos arroja información relevante. Además, los 2 valores son cercanos, y por ende, no permite realizar una diferenciación útil. Se decide eliminarlas.

In [72]:
df_merged3 = df_merged3.drop('PerformanceRating', axis=1)

Para finalizar, se observa entonces que el dataframe final a utilizar en la exploración de datos posee 27 columnas (26 variables explicativas y su respectiva variable objetivo). Este dataframe se exporta a la carpeta 'data' con la finalidad de poderla leer en los archivos siguientes.

In [73]:
df_merged3.head()

EmployeeID  Year  Age     BusinessTravel              Department  \
0           1  2015   51      Travel_Rarely                   Sales   
2           2  2015   31  Travel_Frequently  Research & Development   
3           3  2015   32  Travel_Frequently  Research & Development   
5           4  2015   38         Non-Travel  Research & Development   
7           5  2015   32      Travel_Rarely  Research & Development   

   DistanceFromHome  Education EducationField  JobLevel MaritalStatus  ...  \
0                 6          2  Life Sciences         1       Married  ...   
2                10          1  Life Sciences         1        Single  ...   
3                17          4          Other         4       Married  ...   
5                 2          5  Life Sciences         3       Married  ...   
7                10          1        Medical         1        Single  ...   

   YearsAtCompany  YearsSinceLastPromotion  YearsWithCurrManager  \
0               1                        0                     0   
2               5                        1                     4   
3               5                        0                     3   
5               8                        7                     5   
7               6                        0                     4   

   EnvironmentSatisfaction  JobSatisfaction  WorkLifeBalance  JobInvolvement  \
0                        3                4                2               3   
2                        3                2                4               2   
3                        2                2                1               3   
5                        4                4                3               2   
7                        4                1                3               3   

   retirementType  resignationReason  Attrition  
0  Not Applicable     Not Applicable          0  
2     Resignation             Others          1  
3  Not Applicable     Not Applicable          0  
5  Not Applicable     Not Applicable          0  
7  Not Applicable     Not Applicable          0  

[5 rows x 26 columns]

In [74]:
# Exportación del dataframe para su posterior uso
df_merged3.to_csv('data/df_employees.csv', index=False)